### PR TITLE
Advanced macro expansion

### PR DESCRIPTION
**Pull request by vlad20012**
The epic work begins here

### CODE REVIEW

**Comment by vlad20012** [comment]
Ok, now it can expand `error_chain`

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt [hunk 1]
```diff
@@ -5,46 +5,134 @@
...
+import org.rust.openapiext.Testmark
+import java.util.*
+import java.util.Collections.singletonMap
 
-class MacroExpander(project: Project) {
-    val psiFactory = RsPsiFactory(project)
+private data class MacroSubstitution(
+    val variables: Map<String, String>,
+    val groups: List<List<MacroSubstitution>>
+)
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt]: Let's add some docs about it. Especially about `groups` property

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt [hunk 2]
```diff
@@ -5,46 +5,134 @@
...
+        )
+
+        return substituteMacro(macroExpansion.macroExpansionContents, substWithGlobalVars)
+    }
+
+    private fun findMatchingPattern(
+        def: RsMacroDefinition,
+        call: RsMacroCall
+    ): Pair<RsMacroDefinitionCase, MacroSubstitution>? {
+        val arg = createPsiBuilder(call, call.macroBody ?: return null)
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt]: I still don't like `arg` name - it doesn't contain any useful info

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt [hunk 3]
```diff
@@ -5,46 +5,134 @@
...
+                    start = arg.mark()
+                    null
+                }
+            }
+            ?.firstOrNull()
+    }
+
+    private fun createPsiBuilder(context: PsiElement, text: String): PsiBuilder {
+        val psiMgr = PsiManager.getInstance(project)
+        val result = DummyHolderFactory.createHolder(psiMgr, context)
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt]: Looks like you can inline `result`

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt [hunk 4]
```diff
@@ -57,21 +145,179 @@ class MacroExpander(project: Project) {
...
             }
         }
+}
+
+private class MacroPattern private constructor(
+    val pattern: Sequence<PsiElement>
+) {
+    fun match(arg: PsiBuilder): MacroSubstitution? {
+        return matchPartial(arg)?.let { result ->
+            if (!arg.eof()) {
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt]: But we just checking that we didn't reach the end of file (macro call body)

[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt]: Indeed)

[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/macros/MacroExpander.kt]: Umm. `eof` is more informative in this case, i guess

**Comment by vlad20012** [comment]
bors r=undin

**Comment by alexander-irbis** [comment]
🎉

**Comment by vasily-kirichenko** [comment]
wow!

**Comment by avkonst** [comment]
Has it been packaged to latest rust plugin release? I am using latest plugin version but error_chain expansion still does not work

**Pull request merged**